lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,368 @@
+            std::cout << std::endl;
+        }
+        return 1;    // we now have a token
+    }
+    return 0;    // we didn't find the token
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+size_t BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    FILE *fp = fopen(_dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    size_t formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        return (-1);
+    }
+
+    if (_dataSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        return (-1);
+    }
+    size_t numValues = _dataSize[0] * _dataSize[1] * _dataSize[2];
+
+    int  n = 1;
+    bool systemLittleEndian = *(char *)&n == 1 ? true : false;
+    bool dataLittleEndian = _dataEndian == "LITTLE" ? true : false;
+    bool needSwap = systemLittleEndian != dataLittleEndian ? true : false;
+
+    // Read a "pencil" of data along the X axis, one row at a time
+    size_t count = max[0] - min[0] + 1;
+    for (int k = min[2]; k <= max[2]; k++) {
+        int zOffset = _dataSize[0] * _dataSize[1] * k;
+        for (int j = min[1]; j <= max[1]; j++) {
+            int xOffset = min[0];
+            int yOffset = _dataSize[0] * j;
+            int offset = formatSize * (xOffset + yOffset + zOffset);
+
+            unsigned char readBuffer[count * formatSize];
+
+            if (needSwap) { _swapBytes(readBuffer, formatSize, numValues); }
+
+            fseek(fp, offset, SEEK_SET);
+            size_t rc = fread(readBuffer, formatSize, count, fp);
+
+            if (rc != count) {
+                if (ferror(fp) != 0) {
+                    MyBase::SetErrMsg("Error reading input file");
+                } else {
+                    MyBase::SetErrMsg("Short read on input file");
+                }
+                return -1;
+            }
+
+            if (_dataFormat == DC::XType::INT32) {
```

Comment:
The prompt message claims that data format could be `BYTE` but it doesn't seem you support it here.

I suggest you limit the supported formats to be only float and double to reduce chance of bugs.

Reply:
Thanks for that.  

Fixed, by describing the required formats as INT, FLOAT, or DOUBLE.  No more BYTE.

Reply:
Do you want to delete line 41 which still suggests `byte` ?

Reply:
I don't think so.  I want to keep all of them for completeness, and have the logic designate BYTE and SHORT as XType::INVALID.  This helps me keep track of what data formats exist in BOV, even though we don't support all of them now.

Reply:
Ok.

Reply:
Sure, see the following permalink.

https://github.com/NCAR/VAPOR/blob/08fc797d25c09feef2255ff5301755dc62d4b972/lib/vdc/BOVCollection.cpp#L296